### PR TITLE
[RelayMiner] Handle gzip compression

### PR DESCRIPTION
**Pull request by okdas**
## Summary

MissKitty42 ran into an issue with RelayMiner. Whenever AppGate Server sends relay to their RelayMiner, the response is an unexpected binary sequence instead of json.

This appears to be related to gzip compression: https://discord.com/channels/553741558869131266/1217918754801975398/1222944716828704898

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR. **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
- [ ] **Documentation changes**: `make docusaurus_start`

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and referenced any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: pkg/relayer/proxy/proxy_test.go [hunk 1]
```diff
@@ -516,6 +518,66 @@ func TestRelayerProxy_Relays(t *testing.T) {
 			expectedErrCode: -32000,
 			expectedErrMsg:  "session expired", // Relay rejected by the supplier
 		},
+		// How do I do this? :D
```
Comment (okdas) [inline pkg/relayer/proxy/proxy_test.go:521]: @red-0ne I'm not familiar with relay miner testing suite, what's the best way to test gzip compressed response?

File: pkg/relayer/proxy/relay_builders.go [hunk 2]
```diff
@@ -10,42 +10,45 @@ import (
 
 // newRelayRequest builds a RelayRequest from an http.Request.
 func (sync *synchronousRPCServer) newRelayRequest(request *http.Request) (*types.RelayRequest, error) {
-	requestBz, err := io.ReadAll(request.Body)
+	requestBody, err := io.ReadAll(request.Body)
 	if err != nil {
 		return nil, err
 	}
+	defer request.Body.Close()
```
Comment (okdas) [inline pkg/relayer/proxy/relay_builders.go:17]: @red-0ne Claude 3 mentioned `Body.Close()` never happens, and I also couldn't find that in the code, so I added it here. Could it be a reason for a goroutine leak? Though the tests seem to pass, I want to check in with you to make sure we're not closing the body twice.

File: pkg/relayer/proxy/proxy_test.go (hunk 1 above)
Comment (okdas) [reply pkg/relayer/proxy/proxy_test.go]: Nvm, we decided test will be added in a subsequent pr.

File: pkg/relayer/proxy/relay_builders.go (hunk 2 above)
Comment (Olshansk) [reply pkg/relayer/proxy/relay_builders.go:17]: It seems dangerous to me to to a `.Close` inside of a helper where the request gets passed in.

I feel like we should do it in `ServeHTTP` inside of `pkg/appgateserver/server.go`.

File: pkg/relayer/proxy/relay_builders.go [hunk 3]
```diff
@@ -10,42 +10,45 @@ import (
...
+// newRelayResponse builds a RelayResponse from a response body reader and a SessionHeader.
 // It also signs the RelayResponse and assigns it to RelayResponse.Meta.SupplierSignature.
-// The response's Body is passed directly into the RelayResponse.Payload field.
+// The response body is passed directly into the RelayResponse.Payload field.
 func (sync *synchronousRPCServer) newRelayResponse(
-	response *http.Response,
+	responseBody io.ReadCloser,
 	sessionHeader *sessiontypes.SessionHeader,
 ) (*types.RelayResponse, error) {
+	defer responseBody.Close()
```
Comment (Olshansk) [inline pkg/relayer/proxy/relay_builders.go:37]: Ditto. I don't think we shold be closing things inside of helpers. It creates side effects.

File: pkg/relayer/proxy/synchronous.go [hunk 4]
```diff
@@ -249,14 +250,28 @@ func (sync *synchronousRPCServer) serveHTTP(
...
 	}
 
+	// Check if the response is gzip-compressed
+	// TODO: Add other algorithms, or an alternative would be to switch to http
+	// client that manages all low-level HTTP decisions for us, something like
+	// https://github.com/imroc/req, https://github.com/valyala/fasthttp or
+	// https://github.com/go-resty/resty
+	var responseBody io.ReadCloser = httpResponse.Body
+	if httpResponse.Header.Get("Content-Encoding") == "gzip" {
+		responseBody, err = gzip.NewReader(httpResponse.Body)
```
Comment (Olshansk) [inline pkg/relayer/proxy/synchronous.go:260]: Can you create a helper that takes `httReponse` and returns the correct one? It'll be easier to add new ones in the future.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/proxy/synchronous.go (hunk 4 above)
Comment (okdas) [reply pkg/relayer/proxy/synchronous.go:260]: Great idea!

**Review by Olshansk — APPROVED** [review]
So much cleaner after the changes!

**Pull request merged**